File: battle.py
"""Top Trumps style battle engine for the Minecraft mob cards.

Rules
-----
* A shuffled deck is dealt evenly between the player and the CPU.
* Whoever holds the turn picks one of the six stats on their top card.
* Both top cards are revealed; the higher value on that stat wins.
* The winner puts both cards (plus anything in the pot) on the bottom
  of their deck and picks the stat next round.
* On a tie both cards go into the pot and the same side picks again.
* You win by taking every card. A long stalemate is called on card count.
"""
from __future__ import annotations

import random
import uuid
from typing import Dict, List, Optional

from cards import CARDS, STATS

MAX_ROUNDS = 500

# in-memory store of running games, keyed by game id
GAMES: Dict[str, Dict] = {}


class BattleError(ValueError):
    """Raised for illegal battle moves (bad stat, wrong turn, finished game)."""


def new_game(deck_size: int = 30, rng: Optional[random.Random] = None) -> Dict:
    """Create a game: shuffle the collection, deal `deck_size` cards evenly."""
    if not 4 <= deck_size <= len(CARDS):
        raise BattleError(f"deck size must be between 4 and {len(CARDS)}")
    if deck_size % 2:
        deck_size -= 1  # both sides need equal hands

    rng = rng or random
    deck = [dict(c) for c in rng.sample(CARDS, deck_size)]
    game = {
        "id": uuid.uuid4().hex[:12],
        "player_deck": deck[: deck_size // 2],
        "cpu_deck": deck[deck_size // 2:],
        "pot": [],
        "turn": "player",
        "round": 0,
        "finished": False,
        "winner": None,
        "last_round": None,
    }
    GAMES[game["id"]] = game
    return game


def get_game(game_id: str) -> Optional[Dict]:
    return GAMES.get(game_id)


def cpu_pick_stat(card: Dict, rng: Optional[random.Random] = None) -> str:
    """CPU strategy: lead with the card's strongest stat."""
    best = max(STATS, key=lambda s: card[s])
    return best


def play_round(game: Dict, stat: Optional[str] = None,
               rng: Optional[random.Random] = None) -> Dict:
    """Play one round. `stat` is required on the player's turn and
    ignored on the CPU's turn (the CPU chooses for itself)."""
    if game["finished"]:
        raise BattleError("game is already finished")

    chooser = game["turn"]
    if chooser == "player":
        if stat is None:
            raise BattleError("it is your turn: pick a stat")
        stat = stat.strip().lower()
        if stat not in STATS:
            raise BattleError(f"unknown stat {stat!r}, choose one of {STATS}")
    else:
        stat = cpu_pick_stat(game["cpu_deck"][0], rng)

    player_card = game["player_deck"].pop(0)
    cpu_card = game["cpu_deck"].pop(0)
    game["round"] += 1

    if player_card[stat] > cpu_card[stat]:
        result = "player"
        game["player_deck"] += [player_card, cpu_card] + game["pot"]
        game["pot"] = []
        game["turn"] = "player"
    elif cpu_card[stat] > player_card[stat]:
        result = "cpu"
        game["cpu_deck"] += [cpu_card, player_card] + game["pot"]
        game["pot"] = []
        game["turn"] = "cpu"
    else:
        result = "tie"
        game["pot"] += [player_card, cpu_card]
        # turn stays with the current chooser

    game["last_round"] = {
        "round": game["round"],
        "chooser": chooser,
        "stat": stat,
        "player_card": player_card,
        "cpu_card": cpu_card,
        "result": result,
    }

    _check_end(game)
    return game["last_round"]


def _check_end(game: Dict) -> None:
    player_out = not game["player_deck"]
    cpu_out = not game["cpu_deck"]
    if player_out and cpu_out:
        game["finished"], game["winner"] = True, "draw"
    elif player_out:
        game["finished"], game["winner"] = True, "cpu"
    elif cpu_out:
        game["finished"], game["winner"] = True, "player"
    elif game["round"] >= MAX_ROUNDS:
        # stalemate: call it on cards held
        p, c = len(game["player_deck"]), len(game["cpu_deck"])
        game["finished"] = True
        game["winner"] = "player" if p > c else "cpu" if c > p else "draw"


def public_state(game: Dict) -> Dict:
    """Game state as seen by the player: CPU cards stay face down."""
    state = {
        "id": game["id"],
        "round": game["round"],
        "turn": game["turn"],
        "finished": game["finished"],
        "winner": game["winner"],
        "player_deck_count": len(game["player_deck"]),
        "cpu_deck_count": len(game["cpu_deck"]),
        "pot_count": len(game["pot"]),
        "player_card": game["player_deck"][0] if game["player_deck"] else None,
        "last_round": game["last_round"],
    }
    return state

File: mod/src/main/java/com/jrpetty/mobtrumps/game/Battle.java
package com.jrpetty.mobtrumps.game;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.List;
import java.util.random.RandomGenerator;

/**
 * A Top Trumps battle between the player and the CPU.
 *
 * The deck is dealt evenly. Whoever holds the turn picks a stat on their
 * top card; the higher value wins both cards plus anything in the pot and
 * picks again. Ties send both cards to the pot. Take every card to win.
 */
public final class Battle {

    public enum Side { PLAYER, CPU, NONE }

    /** The outcome of a single resolved round. */
    public record RoundResult(int round, Side chooser, Stat stat,
                              MobCard playerCard, MobCard cpuCard, Side winner) {
    }

    public static final int MAX_ROUNDS = 500;

    private final Deque<MobCard> playerDeck = new ArrayDeque<>();
    private final Deque<MobCard> cpuDeck = new ArrayDeque<>();
    private final List<MobCard> pot = new ArrayList<>();
    private final RandomGenerator random;
    private Difficulty difficulty = Difficulty.NORMAL;
    private Side turn = Side.PLAYER;
    private int round = 0;
    private boolean finished = false;
    private Side winner = Side.NONE;
    /** Who won the coin flip that settled the most recent drawn round. */
    private Side lastCoin = Side.NONE;

    public Battle(int deckSize, RandomGenerator random) {
        this.random = random;
        deckSize = Math.max(4, Math.min(deckSize, MobCards.ALL.size()));
        if (deckSize % 2 != 0) deckSize--;
        List<MobCard> deck = MobCards.shuffledDeck(deckSize, random);
        for (int i = 0; i < deck.size(); i++) {
            (i < deck.size() / 2 ? playerDeck : cpuDeck).addLast(deck.get(i));
        }
    }

    /**
     * Deal explicit hands to each side (used for deck-builder battles and
     * deck-vs-deck duels). Both hands are shuffled; if one side is empty it
     * is filled with a random hand matching the other's size.
     */
    public Battle(List<MobCard> playerHand, List<MobCard> cpuHand, RandomGenerator random) {
        this.random = random;
        List<MobCard> p = new ArrayList<>(playerHand);
        List<MobCard> c = new ArrayList<>(cpuHand);
        int size = Math.max(2, Math.max(p.size(), c.size()));
        if (p.isEmpty()) p = MobCards.shuffledDeck(size, random);
        if (c.isEmpty()) c = MobCards.shuffledDeck(size, random);
        Collections.shuffle(p, java.util.Random.from(random));
        Collections.shuffle(c, java.util.Random.from(random));
        playerDeck.addAll(p);
        cpuDeck.addAll(c);
    }

    public boolean isFinished() {
        return finished;
    }

    public Side getWinner() {
        return winner;
    }

    public Side getTurn() {
        return turn;
    }

    /**
     * The side that won the coin flip settling the last round, or {@link
     * Side#NONE} if that round was not a draw. Reset at the start of every
     * round, so it always describes the round just played.
     */
    public Side lastCoin() {
        return lastCoin;
    }

    public int getRound() {
        return round;
    }

    public int playerCardCount() {
        return playerDeck.size();
    }

    public int cpuCardCount() {
        return cpuDeck.size();
    }

    public int potCount() {
        return pot.size();
    }

    /** The player's current top card (never null while the game runs). */
    public MobCard playerTopCard() {
        return playerDeck.peekFirst();
    }

    /** The CPU/opponent's current top card (used to prompt a human opponent). */
    public MobCard cpuTopCard() {
        return cpuDeck.peekFirst();
    }

    /** Play one round with the stat picked by whoever holds the turn. */
    public RoundResult playRound(Stat stat) {
        if (finished) throw new IllegalStateException("battle is already finished");
        if (stat == null) throw new IllegalArgumentException("a stat must be chosen");

        Side chooser = turn;
        MobCard playerCard = playerDeck.pollFirst();
        MobCard cpuCard = cpuDeck.pollFirst();
        round++;
        lastCoin = Side.NONE; // only a drawn round sets this

        // Rarity is a "lower wins" stat, so compare direction-aware scores
        int playerScore = stat.score(playerCard.stat(stat));
        int cpuScore = stat.score(cpuCard.stat(stat));

        Side roundWinner;
        if (playerScore > cpuScore) {
            roundWinner = Side.PLAYER;
            playerDeck.addLast(playerCard);
            playerDeck.addLast(cpuCard);
            pot.forEach(playerDeck::addLast);
            pot.clear();
            turn = Side.PLAYER;
        } else if (cpuScore > playerScore) {
            roundWinner = Side.CPU;
            cpuDeck.addLast(cpuCard);
            cpuDeck.addLast(playerCard);
            pot.forEach(cpuDeck::addLast);
            pot.clear();
            turn = Side.CPU;
        } else {
            // Tie: both cards go to the pot and nobody has earned the next pick.
            // Letting the chooser simply go again quietly rewards whoever picked
            // the drawn stat, so the turn is settled on a straight coin flip —
            // an even chance for each side. lastCoin lets the UI show it.
            roundWinner = Side.NONE;
            pot.add(playerCard);
            pot.add(cpuCard);
            turn = random.nextBoolean() ? Side.PLAYER : Side.CPU;
            lastCoin = turn;
        }

        checkEnd();
        return new RoundResult(round, chooser, stat, playerCard, cpuCard, roundWinner);
    }

    public void setDifficulty(Difficulty difficulty) {
        if (difficulty != null) this.difficulty = difficulty;
    }

    public Difficulty getDifficulty() {
        return difficulty;
    }

    /** The stat the CPU will pick on its turn, according to the difficulty. */
    public Stat cpuChoice() {
        MobCard top = cpuDeck.peekFirst();
        if (top == null) return Stat.HEALTH;
        return switch (difficulty) {
            case EASY -> Stat.values()[random.nextInt(Stat.values().length)];
            case NORMAL -> top.bestStat();
            case HARD -> hardChoice(top);
        };
    }

    /** Lead with the stat that has the best odds of beating a random card, with a little bluff. */
    private Stat hardChoice(MobCard top) {
        Stat best = Stat.HEALTH;
        Stat second = Stat.HEALTH;
        double bestOdds = -1;
        double secondOdds = -1;
        for (Stat s : Stat.values()) {
            double odds = MobCards.winOdds(s, top.stat(s));
            if (odds > bestOdds) {
                second = best;
                secondOdds = bestOdds;
                best = s;
                bestOdds = odds;
            } else if (odds > secondOdds) {
                second = s;
                secondOdds = odds;
            }
        }
        // if the runner-up is nearly as strong, occasionally bluff with it so a human can't read the CPU
        if (secondOdds >= 0 && bestOdds - secondOdds < 0.15 && random.nextDouble() < 0.25) {
            return second;
        }
        return best;
    }

    private void checkEnd() {
        boolean playerOut = playerDeck.isEmpty();
        boolean cpuOut = cpuDeck.isEmpty();
        if (playerOut && cpuOut) {
            finished = true;
            winner = Side.NONE; // everything ended up in the pot: a draw
        } else if (playerOut) {
            finished = true;
            winner = Side.CPU;
        } else if (cpuOut) {
            finished = true;
            winner = Side.PLAYER;
        } else if (round >= MAX_ROUNDS) {
            finished = true; // stalemate: call it on cards held
            winner = playerDeck.size() > cpuDeck.size() ? Side.PLAYER
                    : cpuDeck.size() > playerDeck.size() ? Side.CPU : Side.NONE;
        }
    }
}
